### DISCUSSION TITLE
Any guide of how to develop a plugin and the stabilize plan?

### DISCUSSION THREAD

**Discussion by lxdlam** [iina/iina#4161] [Q&A] [answered]
Hi community!

I've been reading through the codebase and have some idea to contribute a built-in music scrobble feature for IINA. From searching I discovered some plugin examples and a type definition repository.

But from the codebase, I still have some questions:
1. How the plugin permission subsystem work? If I want to grant the permission that I need to access the network, how to make user approve it?
2. What is supported of the JS runtime? Can I import `axios` or only `fetch` API is supported?
3. How to make it configurable? Where to store the configuation? To prompt my own page or to use IINA built-in settings window to configure?

I think it still require some work ahead to make the plugin system stable, is there any plan of it? Or I may gonna to contribute it using Swift.

**Comment by low-batt** [comment, marked as answer]
There is a lot of discussion about the plugin system in PR #3762, the PR that was used to merge the plugin system into the develop branch.

Last I heard the API documentation was not ready. The documentation must be completed before the plugin system can be enabled by default in a release. The current plan is for that to happen in the next major feature release. At the moment the focus is mostly on fixing defects including some regressions.

Various developers have been trying out the plugin system and submitting PRs with fixes. If you look over commits you will see fixes multiple plugin system issues. You are welcome to try it out and submit plugin related issues if you discover problems.

**Reply by lxdlam** [reply]
Glad to hear that! I'll invesitgate it more and try the best approach. Thanks!
